### PR TITLE
Fix read checks in shell file reader

### PR DESCRIPTION
**Pull request by NicknineTheEagle**
# Description

This fixes a problem in `FileReader::Read` where it was not properly checking if `DOS_ReadFile` has succeeded or not which, in some cases, caused batch script parser to get stuck in infinite loop.
Sample batch script inside archive: [d2_19pat.zip](https://github.com/dosbox-staging/dosbox-staging/files/13220961/d2_19pat.zip)

# Manual testing

Ran PATCH.BAT from the attached sample. Dosbox hanged at the end of the script without the fix and worked fine with the fix.
Tested benchmark pack from Phil's Computer Lab: https://www.philscomputerlab.com/dos-benchmark-pack.html
Tested RUN.BAT from Bubble Bobble from eXo's pack.
Tested these cases:
1. Create a .BAT file that's empty (zero-bytes).
2. Create a .BAT file that contains a single 0 or null characters.
3. Create a .BAT that contains valid lines, but ends with a 0 or null instead of a carriage return.
4. Create a .BAT that contains valid lines but that doesn't end with a null or a CR LF (It just ends at the last valid character).
5. Create a .BAT that ends with CR LF.

No issues noticed.

# Checklist

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by weirddan455** [comment]
This looks like it may be a more widespread issue.  Lots of callers aren't checking the return value.

```
$ git grep DOS_ReadFile
include/dos_inc.h:bool DOS_ReadFile(uint16_t handle,uint8_t * data,uint16_t * amount, bool fcb = false);
src/dos/dos.cpp:                        DOS_ReadFile(STDIN,&c,&n);
src/dos/dos.cpp:                                DOS_ReadFile(STDIN,&c,&n);
src/dos/dos.cpp:                                DOS_ReadFile (STDIN,&c,&n);
src/dos/dos.cpp:                                DOS_ReadFile(STDIN,&c,&n);
src/dos/dos.cpp:                                        DOS_ReadFile(STDIN,&c,&n);
src/dos/dos.cpp:                        if (DOS_ReadFile(reg_bx,dos_copybuf,&toread)) {
src/dos/dos.cpp:                DOS_ReadFile(STDIN, &code, &count);
src/dos/dos_execute.cpp:        if (!DOS_ReadFile(fhandle,(uint8_t *)&head,&len)) {
src/dos/dos_execute.cpp:                                DOS_ReadFile(fhandle,loadbuf,&dataread);
src/dos/dos_execute.cpp:                                DOS_ReadFile(fhandle,loadbuf,&dataread);
src/dos/dos_execute.cpp:                DOS_ReadFile(fhandle,loadbuf,&readsize);
src/dos/dos_execute.cpp:                        readsize=0x8000;DOS_ReadFile(fhandle,loadbuf,&readsize);
src/dos/dos_execute.cpp:                        readsize=(uint16_t)imagesize;DOS_ReadFile(fhandle,loadbuf,&readsize);
src/dos/dos_execute.cpp:                        readsize=4;DOS_ReadFile(fhandle,(uint8_t *)&relocpt,&readsize);
src/dos/dos_files.cpp:bool DOS_ReadFile(uint16_t entry,uint8_t * data,uint16_t * amount,bool fcb) {
src/dos/dos_files.cpp:  if (!DOS_ReadFile(fhandle,dos_copybuf,&toread,true)) return FCB_READ_NODATA;
src/dos/program_intro.cpp:    DOS_ReadFile (STDIN,&c,&n);
src/dos/program_intro.cpp:    DOS_ReadFile (STDIN,&c,&n);
src/dos/program_intro.cpp:    DOS_ReadFile(STDIN, &c, &n);
src/dos/program_more_output.cpp:                DOS_ReadFile(STDIN, &code, &count);
src/dos/program_more_output.cpp:                DOS_ReadFile(STDIN, &tmp, &count);
src/dos/program_more_output.cpp:        DOS_ReadFile(input_handle, reinterpret_cast<uint8_t *>(&code), &count);
src/hardware/reelmagic/driver.cpp:                      if (!DOS_ReadFile(_pspEntry, data, &transactionAmount))
src/shell/file_reader.cpp:      DOS_ReadFile(handle, &data, &bytes_read);
src/shell/shell_cmds.cpp:                                                               DOS_ReadFile(sourceHandle, buffer, &toread);
src/shell/shell_cmds.cpp:               DOS_ReadFile(handle,&c,&n);
src/shell/shell_cmds.cpp:       DOS_ReadFile(STDIN, &c, &n);
src/shell/shell_cmds.cpp:               DOS_ReadFile(STDIN, &c, &n); // read extended key
src/shell/shell_cmds.cpp:               DOS_ReadFile(STDIN, reinterpret_cast<uint8_t *>(&choice), &bytes_read);
src/shell/shell_cmds.cpp:                               if (!DOS_ReadFile(source_handle, buffer, &bytes_requested)) {
src/shell/shell_misc.cpp:               while (!DOS_ReadFile(input_handle, &data, &byte_count)) {
src/shell/shell_misc.cpp:                       DOS_ReadFile(input_handle, &data, &byte_count);
```

I wonder if it's safe to modify `DOS_ReadFile`to set `*amount = 0` on error if that's the thing the callers are actually checking.  It would probably be useful to go through these and see what's up.  In this case `bytes_read` got initialized to 1 by the caller for some reason.

**Comment by NicknineTheEagle** [comment]
https://github.com/dosbox-staging/dosbox-staging/actions/runs/6711865862/job/18240193833?pr=3076
Why did number of warnings go up by 4 here with none of them being in file_reader.cpp?

**Comment by NicknineTheEagle** [comment]
> This looks like it may be a more widespread issue. Lots of callers aren't checking the return value.
> 
> ```
> $ git grep DOS_ReadFile
> include/dos_inc.h:bool DOS_ReadFile(uint16_t handle,uint8_t * data,uint16_t * amount, bool fcb = false);
> src/dos/dos.cpp:                        DOS_ReadFile(STDIN,&c,&n);
> src/dos/dos.cpp:                                DOS_ReadFile(STDIN,&c,&n);
> src/dos/dos.cpp:                                DOS_ReadFile (STDIN,&c,&n);
> src/dos/dos.cpp:                                DOS_ReadFile(STDIN,&c,&n);
> src/dos/dos.cpp:                                        DOS_ReadFile(STDIN,&c,&n);
> src/dos/dos.cpp:                        if (DOS_ReadFile(reg_bx,dos_copybuf,&toread)) {
> src/dos/dos.cpp:                DOS_ReadFile(STDIN, &code, &count);
> src/dos/dos_execute.cpp:        if (!DOS_ReadFile(fhandle,(uint8_t *)&head,&len)) {
> src/dos/dos_execute.cpp:                                DOS_ReadFile(fhandle,loadbuf,&dataread);
> src/dos/dos_execute.cpp:                                DOS_ReadFile(fhandle,loadbuf,&dataread);
> src/dos/dos_execute.cpp:                DOS_ReadFile(fhandle,loadbuf,&readsize);
> src/dos/dos_execute.cpp:                        readsize=0x8000;DOS_ReadFile(fhandle,loadbuf,&readsize);
> src/dos/dos_execute.cpp:                        readsize=(uint16_t)imagesize;DOS_ReadFile(fhandle,loadbuf,&readsize);
> src/dos/dos_execute.cpp:                        readsize=4;DOS_ReadFile(fhandle,(uint8_t *)&relocpt,&readsize);
> src/dos/dos_files.cpp:bool DOS_ReadFile(uint16_t entry,uint8_t * data,uint16_t * amount,bool fcb) {
> src/dos/dos_files.cpp:  if (!DOS_ReadFile(fhandle,dos_copybuf,&toread,true)) return FCB_READ_NODATA;
> src/dos/program_intro.cpp:    DOS_ReadFile (STDIN,&c,&n);
> src/dos/program_intro.cpp:    DOS_ReadFile (STDIN,&c,&n);
> src/dos/program_intro.cpp:    DOS_ReadFile(STDIN, &c, &n);
> src/dos/program_more_output.cpp:                DOS_ReadFile(STDIN, &code, &count);
> src/dos/program_more_output.cpp:                DOS_ReadFile(STDIN, &tmp, &count);
> src/dos/program_more_output.cpp:        DOS_ReadFile(input_handle, reinterpret_cast<uint8_t *>(&code), &count);
> src/hardware/reelmagic/driver.cpp:                      if (!DOS_ReadFile(_pspEntry, data, &transactionAmount))
> src/shell/file_reader.cpp:      DOS_ReadFile(handle, &data, &bytes_read);
> src/shell/shell_cmds.cpp:                                                               DOS_ReadFile(sourceHandle, buffer, &toread);
> src/shell/shell_cmds.cpp:               DOS_ReadFile(handle,&c,&n);
> src/shell/shell_cmds.cpp:       DOS_ReadFile(STDIN, &c, &n);
> src/shell/shell_cmds.cpp:               DOS_ReadFile(STDIN, &c, &n); // read extended key
> src/shell/shell_cmds.cpp:               DOS_ReadFile(STDIN, reinterpret_cast<uint8_t *>(&choice), &bytes_read);
> src/shell/shell_cmds.cpp:                               if (!DOS_ReadFile(source_handle, buffer, &bytes_requested)) {
> src/shell/shell_misc.cpp:               while (!DOS_ReadFile(input_handle, &data, &byte_count)) {
> src/shell/shell_misc.cpp:                       DOS_ReadFile(input_handle, &data, &byte_count);
> ```
> 
> I wonder if it's safe to modify `DOS_ReadFile`to set `*amount = 0` on error if that's the thing the callers are actually checking. It would probably be useful to go through these and see what's up. In this case `bytes_read` got initialized to 1 by the caller for some reason.

Damn, this looks bad. Definitely worth looking into but not sure if this is within the scope of PR or even current release.

**Comment by FeralChild64** [comment]
Sometimes a minor/patch update to MSVC can make it report more warnings than before.

**Comment by johnnovak** [comment]
> Damn, this looks bad. Definitely worth looking into but not sure if this is within the scope of this PR or even the current release.

Agreed. Best to address this after 0.81.0 is out. Should be fixed, but then this bug hasn't caused major issues so far.

**Comment by NicknineTheEagle** [comment]
> Sometimes a minor/patch update to MSVC can make it report more warnings than before.

So what should I do in this case?

**Comment by kcgen** [comment]
I'm not seeing new warnings from this PR, so hopefully it's just a disconnect versus main.

@NicknineTheEagle , can you pull and rebase against `main`?

The last `main` build generated 875 warnings from MSVC:

https://github.com/dosbox-staging/dosbox-staging/actions/runs/6707497783/job/18226174137

**Comment by kcgen** [comment]
This is a useful fix and I think we should try to add it, but the reported testing:

> Ran `PATCH.BAT` from the attached sample. Dosbox hanged at the end of the script without the fix and worked fine with the fix.

Doesn't report any regression testing using other (previously handled) .BAT files.

@NicknineTheEagle -> can you run this PR through a lot more .BAT files and report back?

**Comment by NicknineTheEagle** [comment]
> This is a useful fix and I think we should try to add it, but the reported testing:
> 
> > Ran `PATCH.BAT` from the attached sample. Dosbox hanged at the end of the script without the fix and worked fine with the fix.
> 
> Doesn't report any regression testing using other (previously handled) .BAT files.
> 
> @NicknineTheEagle -> can you run this PR through a lot more .BAT files and report back?

I would if I had any or knew which specific cases to test. Do you have a sample pack?

**Comment by kcgen** [comment]
> > This is a useful fix and I think we should try to add it, but the reported testing:
> > > Ran `PATCH.BAT` from the attached sample. Dosbox hanged at the end of the script without the fix and worked fine with the fix.
> > 
> > 
> > Doesn't report any regression testing using other (previously handled) .BAT files.
> > @NicknineTheEagle -> can you run this PR through a lot more .BAT files and report back?
> 
> I would if I had any or knew which specific cases to test. Do you have a sample pack?

Phil's benchmark pack is launched with `dosbench.bat`, which is an elaborate batch file.

If you have any of the GNU tools available in WSL, you can use this to list all your batch files and sort them by size:

```shell
find . -type f -iname '*.bat' -exec du -h {} + | sort -hr | more
```

Here's Bubble Bobble's from eXo pack. It uses a decently large .BAT file with file copies, config-changes, pause, printing, and three levels of `GOTO`:

[BubbBobb.zip](https://github.com/dosbox-staging/dosbox-staging/files/13230054/BubbBobb.zip)

I've tested all audio device choices and confirmed they work.

**Comment by kcgen** [comment]
Also try it with your own cretions to exercise this specific code.

1. Create a .BAT file that's empty (zero-bytes).
2. Create a .BAT file that contains a single `0` or null characters.
3. Create a .BAT that contains valid lines, but ends with a `0` or null instead of a carriage return.

File: src/shell/file_reader.cpp [hunk 1]
```diff
@@ -38,9 +38,9 @@ std::optional<uint8_t> FileReader::Read()
 {
 	std::uint8_t data        = 0;
 	std::uint16_t bytes_read = 1;
-	DOS_ReadFile(handle, &data, &bytes_read);
 
-	if (bytes_read == 0) {
+	bool result = DOS_ReadFile(handle, &data, &bytes_read);
+	if (!result || bytes_read == 0) {
```
Comment (kcgen) [inline src/shell/file_reader.cpp:43]: The logic itself looks good.

Presumably if `DOS_ReadFile` gives us a negative return code, we shouldn't trust the data (even if some quantity of bytes were read).

**Comment by NicknineTheEagle** [comment]
VS build still fails. What the heck is happening?

**Comment by kcgen** [comment]
Hmm.. even compared the logged MSVC versions from the `main` build:
```
MSBuild version 17.7.2+d6990bcfa for .NET Framework
```

Versus this PR MSVC version:
```
MSBuild version 17.7.2+d6990bcfa for .NET Framework
```

My only guess is that Microsoft has done other updates that aren't reflected in these versions.. or the code lets the compiler "see" and warn about more stuff.

**Comment by kcgen** [comment]
> VS build still fails due to warnings. What the heck is happening?

This other PR <https://github.com/dosbox-staging/dosbox-staging/pull/3078> is also rebases against latest `main` and not showing the extra errors.

So my hunch is there's something about this change that's revealing more warnings in "adjacent" code.

**Comment by weirddan455** [comment]
It's using a newer version of the runner image for this PR (and strangely only this PR, not the 2 that were submitted after this).

```
Runner Image
  Image: windows-2022
  Version: 20231029.1.0
  Included Software: https://github.com/actions/runner-images/blob/win22/20231029.1/images/win/Windows2022-Readme.md
  Image Release: https://github.com/actions/runner-images/releases/tag/win22%2F20231029.1
```

Compared to main and other active PRs:

```
Runner Image
  Image: windows-2022
  Version: 20231024.1.0
  Included Software: https://github.com/actions/runner-images/blob/win22/20231024.1/images/win/Windows2022-Readme.md
  Image Release: https://github.com/actions/runner-images/releases/tag/win22%2F20231024.1
```

**Comment by kcgen** [comment]
Great catch with the runner versions, @weirddan455.

GitHub does gradual deployments, so eventually all new CI runs will bring using this newer versions. 

@NicknineTheEagle , suggest bumping up the warning limit in your PRz as your changes are not the culprit.

**Comment by NicknineTheEagle** [comment]
Run all the suggested batch script tests. No issues noticed. No hangs at the end.

**Review by kcgen — APPROVED** [review]
Ok, let's merge this with regression tests confirmed.
Thanks, @NicknineTheEagle.

**Pull request merged**